sdk/samples/iot/paho_iot_hub_pnp_component_sample.c

```diff
@@ -755,9 +741,192 @@ static void on_message_received(
   }
 }
 
+static void process_twin_message(
+    az_span twin_message_span,
+    az_iot_pnp_client_twin_response_type response_type)
+{
+  az_result rc = az_iot_pnp_client_twin_patch_get_publish_topic(
+      &pnp_client,
+      pnp_mqtt_get_request_id(),
+      publish_message.topic,
+      publish_message.topic_length,
+      NULL);
+  if (az_result_failed(rc))
+  {
+    IOT_SAMPLE_LOG_ERROR("Failed to get the Twin Patch topic: az_result return code 0x%08x.", rc);
+    exit(rc);
+  }
+
+  az_json_reader jr;
+  az_span component_name;
+  az_json_token property_name;
+  az_json_reader property_value;
+  int32_t version = 0;
+  rc = az_json_reader_init(&jr, twin_message_span, NULL);
+  if (az_result_failed(rc))
+  {
+    IOT_SAMPLE_LOG_ERROR("Could not initialize the json reader: az_result return code 0x%08x.", rc);
+    exit(rc);
+  }
+
+  rc = az_iot_pnp_client_twin_get_property_version(&pnp_client, &jr, response_type, &version);
+  if (az_result_failed(rc))
+  {
+    IOT_SAMPLE_LOG_ERROR(
+        "Failed to get the twin property version: az_result return code 0x%08x.", rc);
+    exit(rc);
+  }
+
+  while (az_result_succeeded(
+      rc = az_iot_pnp_client_twin_get_next_component_property(
+          &pnp_client, &jr, response_type, &component_name, &property_name, &property_value)))
+  {
+    if (rc == AZ_OK)
+    {
+      if (az_span_is_content_equal(component_name, thermostat_1_name))
```

Comment:
if (az_span_is_content_equal(component_name, thermostat_1_name) 
          || az_span_is_content_equal(component_name, thermostat_2_name))
      {
        rc = pnp_thermostat_process_property_update(
            &pnp_client,
            &component_name,
            &property_name,
            &property_value,
            version,
            publish_message.payload,
            &publish_message.out_payload);
        if (az_result_succeeded(rc))
        {
          // Send response to the updated property.
          publish_mqtt_message(
              publish_message.topic, publish_message.out_payload, IOT_SAMPLE_MQTT_PUBLISH_QOS);
          IOT_SAMPLE_LOG_SUCCESS("Client sent Temperature Sensor 1 reported property message:");
          IOT_SAMPLE_LOG_AZ_SPAN("Payload:", publish_message.out_payload);

          // Receive the response from the server.
          receive_mqtt_message();
        }
      }
      else ...

Reply:
The `&component_name` you have in your refactor isn't a name but a whole component struct. That would change whether it's `1` or `2`.